### PR TITLE
Fix S4049 FP: Do not raise on methods with attributes that are not allowed on properties

### LINKED ISSUE #3140: Fix S4049 FP: Do not raise on methods with attributes that are not allowed on properties
### Description
The warning S4049 (Properties should be preferred) is raised on methods that are decorated on attributes do not allowed on properties (such as web API controller methods).

### Repro steps
``` C#
public class S4049
{
    private static readonly Random rnd = new Random();

    [NoProperty]
    public int GetRandom() => rnd.Next(); // compliant, the attribute is not allowed on properties.

    [AttributeUsage(AttributeTargets.All ^ AttributeTargets.Property)]
    private class NoPropertyAttribute : Attribute { }
}
```
### Related information
* SonarC# Version 8.3.0.14607

### PR DESCRIPTION
**Pull request by Corniel**
Fix for FP on Consider making method '{0}' a property; when there is any attribute on the method not supported on properties.

Fixes #3140

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs [hunk 1]
```diff
@@ -68,31 +71,31 @@ protected override void Initialize(SonarAnalysisContext context)
...
+        private static bool NameStartsWithGet(IMethodSymbol method)
         {
             var nameParts = method.Name.SplitCamelCaseToWords().ToList();
-
             return nameParts.Count > 1 &&
                    nameParts[0] == "GET";
         }
+
+        private static bool AllAttributesAllowProperties(IMethodSymbol method)
+            => method.GetAttributes()
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:94]: Formatting:
```suggestion
        private static bool AllAttributesAllowProperties(IMethodSymbol method) => 
        method.GetAttributes()
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs [hunk 2]
```diff
@@ -68,31 +71,31 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            .Select(attribute => attribute.AttributeClass)
+            .SelectMany(cls => cls.GetAttributes())
+            .Where(attr => attr.AttributeClass.Is(KnownType.System_AttributeUsageAttribute))
+            .Select(attr => attr.ConstructorArguments[0].Value).Cast<AttributeTargets>()
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:98]: To be on the safe side
```suggestion
            .Select(attr => attr.ConstructorArguments[0].Value)
            .OfType<AttributeTargets>()
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs [hunk 3]
```diff
@@ -68,31 +71,31 @@ protected override void Initialize(SonarAnalysisContext context)
...
-        private bool NameStartsWithGet(IMethodSymbol method)
+        private static bool NameStartsWithGet(IMethodSymbol method)
         {
             var nameParts = method.Name.SplitCamelCaseToWords().ToList();
-
             return nameParts.Count > 1 &&
                    nameParts[0] == "GET";
         }
+
+        private static bool AllAttributesAllowProperties(IMethodSymbol method)
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:93]: It took me some time to understand that `AllAttributes` means only usage attributes. 
```suggestion
        private static bool UsageAttributesAllowProperties(IMethodSymbol method)
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs [hunk 4]
```diff
@@ -68,31 +71,31 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            => method.GetAttributes()
+            .Select(attribute => attribute.AttributeClass)
+            .SelectMany(cls => cls.GetAttributes())
+            .Where(attr => attr.AttributeClass.Is(KnownType.System_AttributeUsageAttribute))
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:97]: Simplify:
```suggestion
            .SelectMany(cls => cls.GetAttributes(KnownType.System_AttributeUsageAttribute))
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:93]: This method does a lot. Can you add test cases to demonstrate it's behavior in various scenarios? For example
* with another attribute that has no explicit usage
* with more attributes, one allowed and one disallowed on property
* with `AttributeTargets.All` to demonstrate that the `HasFlags`

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs [hunk 5]
```diff
@@ -68,31 +71,31 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            }
-
-            return method.Parameters.Length == 0 &&
-                   !method.ReturnsVoid &&
-                   !method.ReturnType.Is(TypeKind.Array) &&
-                   method.Name != "GetEnumerator" &&
-                   method.Name != "GetAwaiter" &&
-                   NameStartsWithGet(method);
-        }
+            => method.Parameters.Length == 0
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:74]: While looking at this method, there are 10 conditions with a code smell. And 8 of them are negative. And there's no nice way out of it.

Can this be improved somehow? While I usually prefer arrow methods, in this case:
- changing `return && ! ...` to `if || ... return false` would remove some noise
- Return type is checked on 3 lines => could be extracted

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Hi @Corniel , Thank you for your contribution! I left some suggestions for improvements.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Hi @Corniel , Thank you for your contribution! I left some suggestions for improvements.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs (hunk 5 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:74]: I though about grouping them in logical subsets, describing their intent. Ill try to improve it.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs [hunk 6]
```diff
@@ -67,32 +70,39 @@ protected override void Initialize(SonarAnalysisContext context)
...
-
-            return nameParts.Count > 1 &&
-                   nameParts[0] == "GET";
-        }
+        private static bool UsageAttributesAllowProperties(IMethodSymbol method) =>
+            method.GetAttributes()
+                .Select(attribute => attribute.AttributeClass)
+                .SelectMany(cls => cls.GetAttributes(KnownType.System_AttributeUsageAttribute))
+                .Select(attr => attr.ConstructorArguments[0].Value)
+                .Cast<AttributeTargets>()
```
Comment (Corniel) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:105]: Has to be a cast. the value of this attribute is an int to start with. Note that an UsageAttribute has no empty/default constructor, so all statements are safe.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs (hunk 3 above)
Comment (Corniel) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs]: I think I covered them all. Just let me know if I missed one.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:105]: Thanks, I missed that

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs (hunk 5 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs]: Nice!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesShouldBePreferred.cs:105]: Let's make it simple then
```suggestion
				.Select(attr => (AttributeTargets)attr.ConstructorArguments[0].Value)
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
One last improvement.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
@Corniel, there's a [code smell](https://sonarcloud.io/project/issues?id=sonaranalyzer-dotnet&open=AXYdyT0uI8aslJC_BHw0&pullRequest=3790&resolved=false&types=CODE_SMELL) about the file containing a tab character. Can you please fix that to spaces?

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource Why is the build failing? I can not see that (I don't have the access).

**Comment by pavel-mikula-sonarsource** [comment]
@Corniel It's probably a random error and restart should fix that. AzurePipelines had some incident this morning and we're getting random failures since then.

**Comment by Corniel** [comment]
> 
> 
> @Corniel, there's a [code smell](https://sonarcloud.io/project/issues?id=sonaranalyzer-dotnet&open=AXYdyT0uI8aslJC_BHw0&pullRequest=3790&resolved=false&types=CODE_SMELL) about the file containing a tab character. Can you please fix that to spaces?

Strange. My defaults are one spaces, the editor-config is on spaces... Anyway, fixed them.

**Comment by pavel-mikula-sonarsource** [comment]
Oh, and it was me adding the tabs :) GitHub UI betrayed me on this one.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Comment by pavel-mikula-sonarsource** [comment]
Error from the build doesn't make any sense in context of this PR:

>[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.7:run (copy-analyzer-data) on project sonar-csharp-plugin: An Ant BuildException has occured: D:\a\1\s\sonaranalyzer-dotnet\src\SonarAnalyzer.RuleDescriptorGenerator\bin\Release\net46\cs does not exist.
[ERROR] around Ant part ...<copy todir="D:\a\1\s\sonar-csharp-plugin\target/analyzer/org/sonar/plugins/csharp">... @ 13:87 in D:\a\1\s\sonar-csharp-plugin\target\antrun\build-main.xml
[ERROR] -> [Help 1]

**Comment by Corniel** [comment]
LOL, not at all indeed.

**Comment by pavel-mikula-sonarsource** [comment]
I'll try to rerun it later. AZP has a bad day today, generating random failures.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel. 
Good news: We have just merged #3793 with the rename change. 
Bad news: Please rebase all your PRs on latest master.

I was trying to push this one before the rename, but unluckily the CI didn't have a good day today.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
Rebase is needed after #3793

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Rebase is needed after #3793 (fix of previous message with wrong status)

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource :(

**Comment by pavel-mikula-sonarsource** [comment]
@Corniel , will you try to rebase this? It's really close to be merged and it's a good time to fix FP, as we're currently on a sprint to kill FPs.

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource Rebase done.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel, something went wrong with this PR and the rebase :( I can see only 4 commits (I believe there were more) and there's only one file with test cases in the modified files.

Can you please take one more look?

**Comment by pavel-mikula-sonarsource** [comment]
We plan to release on Friday/Monday. It would be great to deliver this fix.

**Comment by Corniel** [comment]
Fortunately, I also copied everything before I did a rebase, so I could recover the implementation changes. :)

**Comment by pavel-mikula-sonarsource** [comment]
I usually use `git reflog` and `git checkout <commitid>` to recover broken rebases and lost commits. Useful stuff 🙂

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**